pkg/util/fxutil/fxoptional/component.go

```diff
@@ -0,0 +1,97 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package fxoptional
+
+import "github.com/DataDog/datadog-agent/pkg/util/optional"
+
+// Component is an optional component type.
+// The main purpose of this component is to make it explicit that the component may not be present.
+// Here is an full example of how to use it:
+//
+//	// ------------------------------------------
+//	// Component.go
+//	// ------------------------------------------
+//	type Component fxoptional.Component[MyInterface]
+//
+//	type MyInterface interface {
+//	}
+//
+//	// ------------------------------------------
+//	// Another file implements the component
+//	// ------------------------------------------
+//
+//	// Module defines the fx options for this component.
+//	func Module() fxutil.Module {
+//		return fxutil.Component(
+//			fx.Provide(newMyComponent))
+//	}
+//
+//	type dependencies struct {
+//		fx.In
+//
+//		Config    config.Component
+//		Lifecycle fx.Lifecycle
+//	}
+//
+//	type myComponent struct{}
+//
+//	func newMyComponent(deps dependencies) Component {
+//		if !deps.Config.GetBool("mycomponent.enabled") {
+//			return fxoptional.None[MyInterface]()
+//		}
+//
+//		// Use lifecycle hooks if needed by the component
+//		deps.Lifecycle.Append(fx.Hook{
+//			OnStart: func(context.Context) error {
+//				return nil
+//			},
+//			OnStop: func(context.Context) error {
+//				return nil
+//			},
+//		})
+//		return fxoptional.New[MyInterface](myComponent{})
+//	}
+//
+// // ------------------------------------------
+// // Usage
+// // ------------------------------------------
+//
+//	func UseMyComponent(comp Component) {
+//		myComp, ok := comp.Get()
+//		if ok {
+//			// Do something with myComp
+//		}
+//	}
+//
+// It is recommended to use fxoptional.Component instead of optional.Option[Component].
+// Consider the following example of code
+//
+//	 func run(myComponent Component) {
+//	 }
+//
+//	func MyFunc() {
+//		fxutil.OneShot(run, Module())
+//	}
+//
+// This code fails when using optional.Option[Component] because the type of the component is not Component but optional.Option[Component].
```

Comment:
I'm a bit confused after reading this comment and checking the return value of the `New` and `None` functions from this file. Since the signature of those functions also returns `optional.Option[T]`. Is there any magic happening with Go generics?